### PR TITLE
Deploy Tools using ArgoCD if installed

### PR DESCRIPTION
**Pull request by CloudoguSiebels**

### CODE REVIEW

File: src/main/groovy/com/cloudogu/gitops/features/deployment/ArgoCdApplicationStrategy.groovy [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    private CommandExecutor commandExecutor
+
+    ArgoCdApplicationStrategy(
+            Map config,
+            FileSystemUtils fileSystemUtils,
+            CommandExecutor commandExecutor = new CommandExecutor(),
+            ScmmRepoFactory scmmRepoFactory = null
+    ) {
+        this.fileSystemUtils = fileSystemUtils
+        def defaultFactory = { Map factoryConfig, String factoryRepoTarget, CommandExecutor factoryCommandExecutor -> new ScmmRepo(factoryConfig, factoryRepoTarget, factoryCommandExecutor) } as ScmmRepoFactory
```
Comment (CloudoguSiebels) [inline src/main/groovy/com/cloudogu/gitops/features/deployment/ArgoCdApplicationStrategy.groovy:27]: Apparently this does not work in Graal and need to be reworked.

Comment (schnatterer) [reply src/main/groovy/com/cloudogu/gitops/features/deployment/ArgoCdApplicationStrategy.groovy:27]: For further ref - that's the error that occurs at runtime

```
Exception in thread "main" com.oracle.svm.core.jdk.UnsupportedFeatureError: Proxy class defined by interfaces [interface com.cloudogu.gitops.features.deployment.ArgoCdApplicationStrategy$ScmmRepoFactory] not found. Generating proxy classes at runtime is not supported. Proxy classes need to be defined at image build time by specifying the list of interfaces that they implement. To define proxy classes use -H:DynamicProxyConfigurationFiles=<comma-separated-config-files> and -H:DynamicProxyConfigurationResources=<comma-separated-config-resources> options.
17:06:30  	at org.graalvm.nativeimage.builder/com.oracle.svm.core.util.VMError.unsupportedFeature(VMError.java:89)
17:06:30  	at org.graalvm.nativeimage.builder/com.oracle.svm.core.reflect.proxy.DynamicProxySupport.getProxyClass(DynamicProxySupport.java:171)
17:06:30  	at java.base@17.0.5/java.lang.reflect.Proxy.getProxyConstructor(Proxy.java:47)
17:06:30  	at java.base@17.0.5/java.lang.reflect.Proxy.newProxyInstance(Proxy.java:1037)
17:06:30  	at org.codehaus.groovy.runtime.DefaultGroovyMethods.asType(DefaultGroovyMethods.java:12332)
17:06:30  	at org.codehaus.groovy.runtime.dgm$53.doMethodInvoke(Unknown Source)
17:06:30  	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:274)
17:06:30  	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
17:06:30  	at org.codehaus.groovy.runtime.InvokerHelper.invokePogoMethod(InvokerHelper.java:1029)
17:06:30  	at org.codehaus.groovy.runtime.InvokerHelper.invokeMethod(InvokerHelper.java:1012)
17:06:30  	at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.invokeMethodN(ScriptBytecodeAdapter.java:180)
17:06:30  	at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.asType(ScriptBytecodeAdapter.java:603)
17:06:30  	at com.cloudogu.gitops.features.deployment.ArgoCdApplicationStrategy.<init>(ArgoCdApplicationStrategy.groovy:27)
17:06:30  	at com.cloudogu.gitops.features.deployment.ArgoCdApplicationStrategy.<init>(ArgoCdApplicationStrategy.groovy:23)
17:06:30  	at com.cloudogu.gitops.features.deployment.Deployer.<init>(Deployer.groovy:20)
17:06:30  	at java.base@17.0.5/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
17:06:30  	at java.base@17.0.5/java.lang.reflect.Constructor.newInstance(Constructor.java:480)
17:06:30  	at org.codehaus.groovy.reflection.CachedConstructor.invoke(CachedConstructor.java:72)
17:06:30  	at org.codehaus.groovy.runtime.callsite.ConstructorSite$ConstructorSiteNoUnwrapNoCoerce.callConstructor(ConstructorSite.java:105)
17:06:30  	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallConstructor(CallSiteArray.java:59)
17:06:30  	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:263)
17:06:30  	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:277)
17:06:30  	at com.cloudogu.gitops.features.Mailhog.<init>(Mailhog.groovy:25)
17:06:30  	at com.cloudogu.gitops.Application.registerFeatures(Application.groovy:35)
17:06:30  	at com.cloudogu.gitops.Application.<init>(Application.groovy:18)
17:06:30  	at com.cloudogu.gitops.cli.GitopsPlaygroundCli.run(GitopsPlaygroundCli.groovy:106)
17:06:30  	at picocli.CommandLine.executeUserObject(CommandLine.java:1939)
17:06:30  	at picocli.CommandLine.access$1300(CommandLine.java:145)
17:06:30  	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
17:06:30  	at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
17:06:30  	at picocli.CommandLine$RunLast.handle(CommandLine.java:2314)
17:06:30  	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
17:06:30  	at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
17:06:30  	at picocli.CommandLine.execute(CommandLine.java:2078)
17:06:30  	at io.micronaut.configuration.picocli.PicocliRunner.execute(PicocliRunner.java:202)
17:06:30  	at io.micronaut.configuration.picocli.PicocliRunner.execute(PicocliRunner.java:179)
17:06:30  	at com.cloudogu.gitops.cli.GitopsPlaygroundCliMain.exec(GitopsPlaygroundCliMain.groovy:24)
17:06:30  	at com.cloudogu.gitops.cli.GitopsPlaygroundCliMain.main(GitopsPlaygroundCliMain.groovy:13)
```

Comment (schnatterer) [reply src/main/groovy/com/cloudogu/gitops/features/deployment/ArgoCdApplicationStrategy.groovy:27]: I presume Graal has it's problem with the closure.  I'd try replacing the closure with a method.
If that works with graal, we could just overwrite the method in the test.

**Comment by CloudoguSiebels** [comment]
> Are you alright with the changes?

I fixed a typo. Otherwise happy with it.

> I'd like to simplify HelmValuesConverter.

Yes, absolutely. I didn't realize that this was possible.

> Iterative installation of the Playground. First time with --argocd, next time with --flux will fail. Any ideas for a pragmatic solution?

Is it allowed to install the playground using different features? e.g. first without parameters, next run add `--vault=dev`, next run add `--monitoring`? If not, we could persist the parameters of the initial run and compare them in the next one.
If that is supported however, what do we expect to happen to omitted features? Do we expect argocd to be uninstalled, when we don't pass the `--argocd` flag in the second run (same for all other tools)?  If not, we should probably remove the ArgoCD-Application from the cluster-resource repository when flux is installed. Then ArgoCD should purge them and we install them again with helm. 

In this case, we probably want to add a concept of uninstalling to the strategies. That way, we would also solve that weird state where ArgoCD **and** Helm manage the application at the same time when upgrading from no parameters to `--argocd`.

> Move system to applications/cluster-resources

I would not do that as long as we also support installation via helm. In that case we don't use the cluster-resources repository; and I wouldn't expect them there.

The systems folder was surprising me as well when I first encountered it. I think we *do* want to have a folder that is common to different parts of the playground. I think this could be solved by naming. 
`system` contains data for our tools. We have various names for these: tools, features, system. Can we unify these? Do you have suggestions?

**Comment by schnatterer** [comment]
We discussed and decided upon the issues mentioned above.

I then did a final manual test (successful).

While trying to finish up this PR, I stumbled on a number of issues on our build system:

* Jenkins plugin dependency issues once again, resolved in e1ee67c 
* then cloudogu/gitops-build-lib#34
* and also (which seems to happen a lot recently :thinking: ) `stderr: fatal: unable to access 'http://scmm-scm-manager/scm/repo/argocd/example-apps/': Failed to connect to scmm-scm-manager port 80: Connection timed out`

**Comment by schnatterer** [comment]
I recognized one more issue: #115.
As this can likely be fixed by upgrading prometheus, and we are planning to do this anyway, I created a separate issue to bring this PR to an end.

**Pull request merged**